python/ray/external_storage.py

```diff
@@ -90,6 +129,223 @@ def restore_spilled_objects(self, keys):
+
+        self.bucket_name = bucket_name
+        self.prefix = prefix
+        self.override_transport_params = override_transport_params or {}
+        self.track_tail_latency = track_tail_latency
+
+        self.session = boto3.Session()
+        self.transport_params = {
+            "session": self.session
+        }.update(self.override_transport_params)
+
+        # --validation--
+        s3 = self.session.resource("s3")
+        # Make sure the bucket exists.
+        bucket = s3.Bucket(self.bucket_name)
+        if not bucket.creation_date:
+            raise ExternalStorageConfigInvalidError(
+                f"Bucket name {self.bucket_name} doesn't "
+                "exist or is not reachable.")
+
+        # --stats--
+        # SANG-TODO Implement it for benchmark.
+        self.download_count = 0
+        self.upload_count = 0
+        self.download_avg_latency = 0
+        self.upload_avg_latency = 0
+
+        if self.track_tail_latency:
+            self.download_latencies = []
+            self.upload_latencies = []
+
+    def spill_objects(self, object_refs):
+        keys = []
+        ray_object_pairs = self._get_objects_from_store(object_refs)
+        for ref, (buf, metadata) in zip(object_refs, ray_object_pairs):
+            key = self.prefix + ref.hex()
+            self._spill_object(key, ref, buf, metadata)
+            keys.append(key.encode())
+        return keys
+
+    def restore_spilled_objects(self, keys):
+        for k in keys:
+            key = k.decode()
+            ref = ray.ObjectRef(bytes.fromhex(key[len(self.prefix):]))
+            self._restore_spilled_object(key, ref)
+
+    def get_stats(self):
+        # TODO(sang): Implement it for benchmark.
+        return {}
+
+    def _spill_object(self, key, ref, buf, metadata):
+        from smart_open import open
+        url = self._build_s3_url(key)
+        with open(
+                url, "wb",
+                transport_params=self.transport_params) as file_like:
+            metadata_len = len(metadata)
+            buf_len = len(buf)
+            file_like.write(metadata_len.to_bytes(8, byteorder="little"))
+            file_like.write(buf_len.to_bytes(8, byteorder="little"))
+            file_like.write(metadata)
+            file_like.write(memoryview(buf))
+
+    def _restore_spilled_object(self, key, ref):
+        from smart_open import open
+        url = self._build_s3_url(key)
+        with open(
+                url, "rb",
+                transport_params=self.transport_params) as file_like:
+            metadata_len = int.from_bytes(
+                file_like.read(8), byteorder="little")
+            buf_len = int.from_bytes(file_like.read(8), byteorder="little")
+            metadata = file_like.read(metadata_len)
+            # read remaining data to our buffer
+            self._put_object_to_store(metadata, buf_len, file_like, ref)
+
+    def _build_s3_url(self, key):
+        return f"s3://{self.bucket_name}/{key}"
```

Comment:
Instead of bucket name and key, the user can provide a storage prefix, e.g. "s3://bucket/foo/"